syntax = "proto3";

package hub;

option go_package = "deepapp_golang_grpc_hub/internal/proto";
option java_package = "com.deepapp.hub";

service HubService {
  rpc Connect(stream Message) returns (stream Message);
}

message Message {
  string id = 1;
  string from = 2;
  string to = 3;
  string channel = 4;
  string content = 5;
  string timestamp = 6;
  MessageType type = 7;
  string action = 8; // For dynamic routing: "register", "request", "response", etc
  map<string, string> metadata = 9; // Flexible metadata
}

enum MessageType {
  DIRECT = 0;
  BROADCAST = 1;
  CHANNEL = 2;
  REGISTER = 3; // Worker registration
  REQUEST = 4;  // Service request
  RESPONSE = 5; // Service response
  WORKER_CALL = 6; // Worker-to-Worker call
}

// Worker registration message
message WorkerRegistration {
  string worker_id = 1;
  string worker_type = 2;
  repeated ServiceCapability capabilities = 3;
  map<string, string> metadata = 4;
}

// Service capability definition
message ServiceCapability {
  string name = 1;
  string description = 2;
  string input_schema = 3; // JSON schema
  string output_schema = 4; // JSON schema
}

message Request {
  RequestType type = 1;
  string data = 2;
}

enum RequestType {
  JSON = 0;
  FILE = 1;
  CONTROL = 2;
}

message Response {
  Status status = 1;
  string data = 2;
}

enum Status {
  OK = 0;
  ERROR = 1;
}